Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift

```diff
@@ -401,13 +405,38 @@ extension HTTPConnectionPool {
             self.failedConsecutiveConnectionAttempts += 1
             self.lastConnectFailure = error
 
+            guard self.retryConnectionEstablishment else {
+                guard let (index, _) = self.connections.failConnection(connectionID) else {
+                    preconditionFailure("Failed to create a connection that is unknown to us?")
+                }
+                self.connections.removeConnection(at: index)
+
+                return .init(
+                    request: self.failAllRequests(reason: error),
+                    connection: .none
+                )
+            }
+
             let eventLoop = self.connections.backoffNextConnectionAttempt(connectionID)
             let backoff = calculateBackoff(failedAttempt: self.failedConsecutiveConnectionAttempts)
             return .init(request: .none, connection: .scheduleBackoffTimer(connectionID, backoff: backoff, on: eventLoop))
         }
 
         mutating func waitingForConnectivity(_ error: Error, connectionID: Connection.ID) -> Action {
             self.lastConnectFailure = error
+
+            guard self.retryConnectionEstablishment else {
+                guard let (index, _) = self.connections.failConnection(connectionID) else {
+                    preconditionFailure("Failed to create a connection that is unknown to us?")
+                }
+                _ = self.connections.closeConnection(at: index)
+
+                return .init(
+                    request: self.failAllRequests(reason: error),
+                    connection: .none
+                )
+            }
+
```

Comment:
As stated before, we should never reach this point, as we never scheduled a backoff timer.

Reply:
I think, if we don't want to wait for connection establishment, we should never wait for connectivity in NW. I think I would prefer a precondition with a comment here.